exploration/0475-overriding-extending-namespacing.md

```diff
@@ -0,0 +1,240 @@
+
+- Developers want to import 3rd party formatting packages and use the package's
+  features from within messages.
+
+- Developers wish to import two or more formatting packages
+  and these might have the same-named functions.
+  For example, there might be both an HTML `p` and TTS `p`
+  function.
+
+- Translators and tools would like a machine-readable way to find out the names
+  and option values for add-on packages.
+
+## Requirements
+
+_What properties does the solution have to manifest to enable the use-cases above?_
+
+- Developers must be able to write functions that do not later collide with items in the default registry.
+- Developers must be able to write function add-ons that do not later collide with items in the default registry.
+- Users should be able to tell visually when an add-on feature has been used vs. a built-in
+- Users should be able to resolve conflicts between add-on packages that use the same
+  function names without altering add-on packages
+-
+
+## Constraints
+
+_What prior decisions and existing conditions limit the possible design?_
+
+- A syntactical prefix must not collide with `nmtoken`, to avoid parsing ambiguities with unquoted literals...
+
+## Proposed Design
+
+_Describe the proposed solution. Consider syntax, formatting, errors, registry, tooling, interchange._
+
+The actual addition and provisioning of features to an implementation is implementation specific.
+Implementations are not required to read the registry format defined by MFv2
+or use it for any particular purpose.
+
+> For example, a Java implementation might use the `ServiceProvider` interface to load
+> functionality, while a Node application might use `requires`.
+
+If an implementation supports user-installed formatters, selectors, function options,
+or expression annotations, it must also support providing "namespace" prefixes for
+each installed set of functionality.
+
+In this design, each namespace prefix is a short string and is associated with a URL.
+The URL is intended to point to some publically-available copy of the add-on library's
+registry description, for use by tooling and as a reference to users such as translators.
+
+There is no requirement that an implementation read the document at the end of the URL,
+nor validate the contents in general or against the actual functionality installed.
+
+> [!NOTE]
+> It is a good idea to use ASCII strings for namespace identifiers.
+> Remember that translators (and others) in many different languages and
+> with many different keyboards need to be able to enter the prefix.
+
+The namespace prefix is part of the `name` production.
+The prefix is limited to eight characters in length and MUST be at least two characters
+in length.
+The prefix is separated from the name by a colon (U+003A COLON).
+The namespace prefix `std` is reserved and refers to the default registry.
+
+```abnf
+name      = [namespace] name-body
+namespace = name-start *7(name-char) namespace-sep
```

Comment:
By using `name-char` here we're allowing namespace names themselves to contain colons. I think we need to split `name-char` and `nmtoken-char`.

Reply:
Added a note. We still have the Nmtoken problem to solve, so called that out in text.

Reply:
Thanks, the note looks good to me.